feat: add ADR-034 struct tags (#6)
birdie-ai/jsonspec: README.md

The library will automatically convert between the different conventions for field names, for
example turning `AuthMethod` into `auth_method`.

You can generate a spec for this type as follows:

    spec, err := jsonspec.For(new(DatabaseSource))

Now you can call `spec.ValidateJSON` to check if a JSON document matches the spec. It'll return an
error if any of the fields have the wrong type or if any fields marked as required are missing.

You can also call `LoadJSON` to load a DatabaseSource from JSON:

    var source DatabaseSource
    err := jsonspec.LoadJSON(data, &source) // data is a []byte

This will generate a spec for DatabaseSource, validate that the input matches the spec, and store
the data in `source`.

## Struct tags

The following tags are recognized on struct fields.

Tags that affect input validation:

- `required:"true"`: the field must be provided in the JSON as a non-zero value.
- `enum:"a,b,c"`: list of allowed values. Only enforced for string fields.

Tags that only annotate the generated spec for downstream tooling (such as the connector
management frontend) and are ignored by `Validate`:

- `description:"..."`: human-readable description of the field.

> - `default:"..."`: default value for the field (parsed according to its type).

Reviewer: ```suggestion
- `default:"..."`: default value for the field (parsed according to its type) if no value is provided in the input JSON.
```

> - `tags:"a,b,c"`: list of free-form tags attached to the field.

Reviewer: After this passes, worth checking if we use this any where for anything. If it's not used, I'd remove it from json spec.

Reviewer: Also didn't we have a secret tag?

Author: tags:"secret" still works the same way. ADR-034 proposes replacing it with config:"credential", which carries the same intent and lets downstream tooling treat the credential layer uniformly. I am keeping both for now so we can migrate connectors one at a time. We will come back here to drop the tags once the rollout is finished

in `person`.

## Struct tags

The following tags are recognized on struct fields:

> - `description:"..."`: human-readable description of the field.

Reviewer: Maybe group the tags in to groups:

- One for generating the json_spec
- One for input data validation

> - `required:"true"`: the field must be set for the object to be valid.

Reviewer: ```suggestion
- `required:"true"`: the field must be provided in the JSON as a non-zero value.
```

- `default:"..."`: default value for the field (parsed according to its type).
- `tags:"a,b,c"`: list of free-form tags attached to the field.
- `enum:"a,b,c"`: list of allowed values. Only enforced for string fields.
- `config:"credential|connection|schedule"`: configuration layer the field
  belongs to. Empty means the field is not user-facing.

> - `advanced:"true"`: field should be hidden behind an "advanced" toggle.

Reviewer: ```suggestion
- `advanced:"true"`: field should be hidden behind an "advanced" toggle in the frontend for connector management.
```

> - `visible_when:"<field>=<value>"`: field only renders when the referenced

Reviewer: ```suggestion
- `visible_when:"<field>=<value>"`: field only renders in the frontend for connector management when the referenced
```

>   sibling matches the given value.

Reviewer: ```suggestion
  field matches the given value.
```

> Here's a connector configuration that exercises the new tags:

Reviewer: The idea isn't to just document whats new.

It's to create a single reference for the behaviours.

Users of the library don't know what's new and old

You don't expect to read the Go docs for a library and see it mention that in v 0.1.bolinha a parameter was added to a function to do x.

You just expect it to explain all of the function parameters.

>     type GitHubSource struct {

Reviewer: This didn't even need to be a new example. You could just extend the example above, and add a small section explaining why we have some tags that do nothing.

Author: replaced the Person example with DatabaseSource so a single example exercises every tag, and the grouping above makes explicit which tags validate ignores
